### PR TITLE
Fixes issue #1021

### PR DESCRIPTION
**Pull request by greglandrum**
This changes the way ring stereochemistry is identified and renders it insensitive to the details of the ring-finding algorithm.

The new implementation is about the same speed as the old one (maybe a bit quicker). It could almost certainly be optimized, but I want to get the fix in first and deal with that later when time arises.

It's been through valgrind and is clean.

### CODE REVIEW

**Comment by bp-kelley** [comment]
This is going to take me a bit to review :)  Nice work.

File: Code/GraphMol/Chirality.cpp [hunk 1]
```diff
@@ -451,65 +451,141 @@ bool atomIsCandidateForRingStereochem(const ROMol &mol, const Atom *atom) {
...
+    if (atomsSeen[atom->getIdx()]) continue;
+    if (atom->getChiralTag() == Atom::CHI_UNSPECIFIED ||
+        atom->hasProp(common_properties::_CIPCode) ||
+        !mol.getRingInfo()->numAtomRings(atom->getIdx()) ||
+        !atomIsCandidateForRingStereochem(mol, atom)) {
+      continue;
+    }
+    // do a BFS from this ring atom along ring bonds and find other
+    // stereochemistry candidates.
+    std::list<const Atom *> nextAtoms;
```
Comment (bp-kelley) [inline Code/GraphMol/Chirality.cpp]: vector would probably be faster.  Better cache coherence.

Comment (bp-kelley) [inline Code/GraphMol/Chirality.cpp]: we can worry about this later, .pop_front can be a pain.

Comment (rvianello) [inline Code/GraphMol/Chirality.cpp]: I vote for a deque then ;-)

File: Code/GraphMol/Chirality.cpp [hunk 2]
```diff
@@ -451,65 +451,141 @@ bool atomIsCandidateForRingStereochem(const ROMol &mol, const Atom *atom) {
...
+      // because we're only going to hit each ring atom once, the first atom we
+      // encounter in a ring is going to end up with all the other atoms set as
+      // stereoAtoms, but each of them will only have the first atom present. We
+      // need to fix that. because the traverse from the first atom only
+      // followed ring bonds, these things are all by definition in one ring
+      // system. (Q: is this true if there's a spiro center in there?)
+      INT_VECT same(mol.getNumAtoms(), 0);
+      BOOST_FOREACH (int ringAtomEntry, ringStereoAtoms) {
+        int ringAtomIdx =
+            ringAtomEntry < 0 ? -ringAtomEntry - 1 : ringAtomEntry - 1;
```
Comment (bp-kelley) [inline Code/GraphMol/Chirality.cpp]: Perhaps use the same tagging logic as above

```
  same * (ratom->getIdx() + 1) : -1
```

Easier on me mentally :)

File: Code/GraphMol/Chirality.cpp (hunk 1 above)
Comment (greglandrum) [inline Code/GraphMol/Chirality.cpp]: can we please do the optimization stuff later? I'm concerned that this is big and complex enough that we'll both end up spending weeks on the review if we convolve correctness and performance (which is no worse than what was there before)

File: Code/GraphMol/molopstest.cpp [hunk 3]
```diff
@@ -4643,6 +4643,7 @@ void _renumberTest(const ROMol *m) {
     }
 
     std::string nSmi = MolToSmiles(*nm, true);
+    if (nSmi != refSmi) std::cerr << refSmi << std::endl << nSmi << std::endl;
```
Comment (bp-kelley) [inline Code/GraphMol/molopstest.cpp]: debugging code?

Comment (greglandrum) [inline Code/GraphMol/molopstest.cpp]: Yeah, it's one of those "if it fails you're going to want to see this anyway, so might as well go ahead and show it"

This pattern is frequent enough that you'd think I'd add something like TEST_EQUAL() or switch to an actual testing framework. ;-)

File: Code/GraphMol/Chirality.cpp [hunk 4]
```diff
@@ -451,65 +451,141 @@ bool atomIsCandidateForRingStereochem(const ROMol &mol, const Atom *atom) {
...
+  boost::dynamic_bitset<> atomsSeen(mol.getNumAtoms());
+  boost::dynamic_bitset<> atomsUsed(mol.getNumAtoms());
+  boost::dynamic_bitset<> bondsSeen(mol.getNumBonds());
+
+  for (ROMol::AtomIterator ait = mol.beginAtoms(); ait != mol.endAtoms();
+       ++ait) {
+    const Atom *atom = *ait;
+    if (atomsSeen[atom->getIdx()]) continue;
+    if (atom->getChiralTag() == Atom::CHI_UNSPECIFIED ||
+        atom->hasProp(common_properties::_CIPCode) ||
```
Comment (bp-kelley) [inline Code/GraphMol/Chirality.cpp]: Is this shorthand for checking for CHI_TETRAHEDRAL_CCW/CW?

I'm not understanding the the difference between chiral and unspecified or chiral and specified here.

File: Code/GraphMol/molopstest.cpp (hunk 3 above)
Comment (bp-kelley) [inline Code/GraphMol/molopstest.cpp]: perhaps put in the BOOST_LOG(rdError)?

File: Code/GraphMol/Chirality.cpp (hunk 4 above)
Comment (greglandrum) [inline Code/GraphMol/Chirality.cpp]: The reason ring stereochemistry is hard (in the current representation) is that the atoms don't have CIP chirality but do have (and need to have) the stereochemistry specified. That's what this is testing for: the stereochemistry is specified and there's no CIP code on the atom.

File: Code/GraphMol/Chirality.cpp (hunk 2 above)
Comment (greglandrum) [inline Code/GraphMol/Chirality.cpp]: afraid I don't understand the suggestion.

Comment (bp-kelley) [inline Code/GraphMol/Chirality.cpp]: Probably solved by a comment, you making indices negative when they are not the same here.    The maths have different signs for unpacking which I found correct but confusing.

File: Code/GraphMol/Chirality.cpp [hunk 5]
```diff
@@ -451,65 +451,141 @@ bool atomIsCandidateForRingStereochem(const ROMol &mol, const Atom *atom) {
...
+
+    while (!nextAtoms.empty()) {
+      const Atom *ratom = nextAtoms.front();
+      nextAtoms.pop_front();
+      atomsSeen.set(ratom->getIdx());
+      if (ratom->getChiralTag() != Atom::CHI_UNSPECIFIED &&
+          !ratom->hasProp(common_properties::_CIPCode) &&
+          atomIsCandidateForRingStereochem(mol, ratom)) {
+        int same = (ratom->getChiralTag() == atom->getChiralTag()) ? 1 : -1;
+        ringStereoAtoms.push_back(same * (ratom->getIdx() + 1));
```
Comment (bp-kelley) [inline Code/GraphMol/Chirality.cpp]: As far as I can tell, this tagging is never actually used, i.e. you always push the tagged index but never actually use the fact that the chiral tags are the same or different.  All I can find is unpacking the tag to the proper atom index.

File: Code/GraphMol/Chirality.cpp [hunk 6]
```diff
@@ -451,65 +451,141 @@ bool atomIsCandidateForRingStereochem(const ROMol &mol, const Atom *atom) {
...
+        INT_VECT lringatoms(0);
+        mol.getAtomWithIdx(ringAtomIdx)
+            ->getPropIfPresent(common_properties::_ringStereoAtoms, lringatoms);
+        CHECK_INVARIANT(lringatoms.size() > 0, "no other ring atoms found.");
+        for (INT_VECT_CI orae = rae + 1; orae != ringStereoAtoms.end();
+             ++orae) {
+          int oringAtomEntry = *orae;
+          int oringAtomIdx =
+              oringAtomEntry < 0 ? -oringAtomEntry - 1 : oringAtomEntry - 1;
+          int theseDifferent = (ringAtomEntry < 0) ^ (oringAtomEntry < 0);
```
Comment (bp-kelley) [inline Code/GraphMol/Chirality.cpp]: Nope, it's used here.

File: Code/GraphMol/Chirality.cpp (hunk 2 above)
Comment (bp-kelley) [inline Code/GraphMol/Chirality.cpp]: When tagging atoms for special purposes, I tend to write inline functions to help me cognitively.  This is a style thing so I don't expect agreement.  Here is an example, however:
```
inline int tagIdxWithSameChirality(unsigned int atomIdx, bool sameChirality) {
  return ((int)sameChirality * rdcast<int>(atomIdx + 1));
}

inline int getIdxFromTax(int tag) {
  return tag > 0 ? tag - 1 : -(tag - 1) 
}

inline bool hasSameChirality(int tag) {
   return tag > 0;
}
```

And here is the confusing usage:

```
BOOST_FOREACH (int ringAtomEntry, ringStereoAtoms) {
        same[getIdxFromTag[ringAtomEntry]] = hasSameChirality(ringAtomEntry);
      }
```

Which I find reads a lot better and should have the same performance.  Just my 2cents though.  Now that I understand the code, It seems to solve the current issue.

**Comment by bp-kelley** [comment]
Ok, I believe the code is correct and am ready to merge.  I might do a small clean-up pass at some point perhaps just a comment line or two  if that's ok, might not happen since I have the big int-keys merger to do before the UGM!

Is there any last bit you want me to check?

**Comment by greglandrum** [comment]
Seems like it's ready to merge.

**Pull request merged**